plugins/hls-code-range-plugin/src/Ide/Plugin/CodeRange.hs

```diff
@@ -126,6 +148,61 @@ findPosition pos root = go Nothing root
             startOfRight <- _start . _codeRange_range <$> V.headM right
             if pos < startOfRight then binarySearchPos left else binarySearchPos right
 
+-- | Traverses through the code range and it children to a folding ranges.
+--
+-- It starts with the root node, converts that into a folding range then moves towards the children.
+-- It converts each child of each root node and parses it to folding range and moves to its children.
+findFoldingRanges :: CodeRange -> [FoldingRange]
+findFoldingRanges r@(CodeRange _ children _) =
+    let frChildren :: [FoldingRange] = concat $ V.toList $ fmap findFoldingRanges children
+    in case createFoldingRange r of
+        Just x  -> x:frChildren
+        Nothing -> []
+
+-- | Parses code range to folding range
+createFoldingRange :: CodeRange -> Maybe FoldingRange
+createFoldingRange (CodeRange (Range (Position lineStart charStart) (Position lineEnd charEnd)) _ ck) = do
+    -- Type conversion of codeRangeKind to FoldingRangeKind
+    let frk = crkToFrk ck
+
+    -- Filtering code ranges that start and end on the same line as need/can not be folded.
+    --
+    -- Eg. A single line function will also generate a Folding Range but it cannot be folded
+    -- because it is already single line, so omiting it.
+    if lineStart == lineEnd
+        then Nothing
+    else case frk of
+        Just _  -> Just (FoldingRange lineStart (Just charStart) lineEnd (Just charEnd) frk)
+        Nothing -> Nothing
+
+-- | Removes all small foldings that start from the same line.
+--
+-- As we are converting nodes of the ast into folding ranges, there are multiple nodes starting from a single line.
+-- A single line of code doesn't mean a single node in AST, so this function removes all the nodes that have a duplicate
+-- start line, ie. they start from the same line.
+--
+-- This function preserves the largest folding range from the ranges that start from the same line.
```

Comment:
I think this is the most important line. This is the property that you want: for each line, we only return one folding range, and it should be the largest amongst all the folding ranges we have.

If we specify the behaviour like that then it suggests other ways we could implement this function, e.g. go through the list maintaining a map from lines to the largest folding range so far that starts on that line, and then each time we see a range, check if it is bigger than the existing one. That would remove the way the current implementation depends on the order.

It also raises some questions: folding ranges can overlap, with neither being bigger than the other (e.g. two ranges on the same line with character ranges [0, 5] and [1,6]). Which one should we pick then?

Lastly, are we sure this is what we want? Your example is interesting: why do we assume that people will want to fold the function rather than the if expression? I wonder if we *can* leave this up to the client. It seems like the server's job should just be to report all the ranges it knows about, and then the client can potentially make the decision about which ones to show to the user. What happens in e.g. vscode if we return multiple folding ranges starting on the same line?

Reply:
> folding ranges can overlap

folding ranges correspond to the ranges of AST nodes. I can not come up with a case where two AST nodes overlap, while neither of them is the parent of the other.

Reply:
Good point! So that gives us some information. We can have parents and children on the same line, though so that suggests that we know the following: code ranges only overlap if one of them entirely contains the other one. That's useful. We still can't totally order them because we could also have *non*-overlapping code ranges on the same line (which one is bigger?).

Reply:
Closing as this functionality has been left for the client to handle